### PR TITLE
Implement repeating sections

### PR DESCRIPTION
**Pull request by MebinAbraham**
### What is the context of this PR?
Added the ability to repeat a section for a given list.

#### Main changes this PR introduces:
- `ProgressStore` is now stored and mapped in the same way as `AnswerStore`, this was done to support `list_item_id`
- `ProgressStore` now stores `Progress` objects. Same way `AnswerStore` stores `Answer` objects.
- Section id has been introduced to a Location object to easily track the current section. Many places fetch the `section_id` from block id, this is now not necessary. This change doesn't remove all previous fetches from the schema.
- Relationship Location object's `from_list_item_id` is now `list_item_id`.
- The `Location`/`RelationshipLocation` object is only created once we know it is a valid URL.
- This PR doesn't add support to display repeating sections on summary pages.

To Do:
* [x]  Fix breaking tests
* [x]  Add unit tests
* [x]  Add functional tests
* [x]  Update validator.

### How to review 
Ensure:
- You are able to repeat a section without any unexpected behaviours.
- Tests scenarios with the primary person, list collectors and relationships etc.

### CODE REVIEW

File: app/data_model/progress_store.py
Comment (MebinAbraham) [inline app/data_model/progress_store.py]:
```suggestion
        """ Builds the progress_store's data structure from a list of progress dictionaries"""
```

File: app/views/handlers/primary_person_list_collector.py
Comment (MebinAbraham) [inline app/views/handlers/primary_person_list_collector.py:39]: This has been removed because if you answer `Yes` to the primary person question, then click Previous, then repost the `Yes` answer, then not submit the name question, the primary person list collector block is never marked as complete which can cause routing issues.

This line was added to fix another issue, where you answered `No` but then changed the answer to `Yes`.

However, the latter is much less likely to happen therefore, for now, after speaking with @ajmaddaford and @bitdivision, I've decided to make it less broken?

Happy to discuss in IRL if the above is unclear.

**Comment by MebinAbraham** [comment]
[Implement repeating sections (M)](https://trello.com/c/aoUVtpy8/2800-implement-repeating-sections-m)

File: app/libs/utils.py
Comment (MebinAbraham) [inline app/libs/utils.py]: Feel free to suggest what this method should be called and where this should live?. Maybe a new `answer_store_helper.py`?

File: app/questionnaire/path_finder.py
Comment (MebinAbraham) [inline app/questionnaire/path_finder.py:89]: A location object is needed since the location comparison in when rules got introduced.

**Comment by MebinAbraham** [comment]
I'm in the process of adding unit/integration tests, its just time consuming, but everything else can be reviewed.

File: app/data_model/progress_store.py
Comment (iwootten) [inline app/data_model/progress_store.py]: This probably should be named differently, as it isn't actually a location

File: app/libs/utils.py
Comment (iwootten) [reply app/libs/utils.py]: This whole file/package would be better merged into existing more meaningful class/packages I think.

File: app/questionnaire/questionnaire_store_updater.py
Comment (iwootten) [inline app/questionnaire/questionnaire_store_updater.py]: The location here could probably be made a normal arg now.

File: app/questionnaire/router.py
Comment (iwootten) [inline app/questionnaire/router.py]: Again, might be better to name this something different..

File: app/questionnaire/router.py
Comment (iwootten) [inline app/questionnaire/router.py]:
```suggestion
            section_id, list_item_id = incomplete_section_locations[0]
```

File: app/views/contexts/summary_context.py
Comment (iwootten) [inline app/views/contexts/summary_context.py]: section_schemas might be a better name for the collection here

File: tests/functional/spec/list_collector_primary_person.spec.js
Comment (iwootten) [inline tests/functional/spec/list_collector_primary_person.spec.js:18]: This test is currently skipped.

**Review by iwootten — CHANGES_REQUESTED** [review]

File: tests/functional/spec/list_collector_primary_person.spec.js
Comment (MebinAbraham) [reply tests/functional/spec/list_collector_primary_person.spec.js:18]: I could remove it or leave it in for reference?. It is related to my comment here:
https://github.com/ONSdigital/eq-survey-runner/pull/2248#discussion_r309666515

File: app/views/contexts/summary_context.py
Comment (MebinAbraham) [reply app/views/contexts/summary_context.py]: I've removed this from this file since I am not adding support for repeating sections on the summary in this PR.

File: app/data_model/progress_store.py
Comment (MebinAbraham) [reply app/data_model/progress_store.py]: In other places, we use `key` but that seems too vague for this as it is used in multiple files, therefore I settled for `section_key`. Feel free to suggest alternatives.

File: app/questionnaire/questionnaire_store_updater.py
Comment (MebinAbraham) [reply app/questionnaire/questionnaire_store_updater.py]: Done.

File: app/questionnaire/router.py
Comment (MebinAbraham) [reply app/questionnaire/router.py]: Same as above.

**Comment by MebinAbraham** [comment]
@iwootten Regarding, https://github.com/ONSdigital/eq-survey-runner/pull/2248#discussion_r312809124

Any suggestions on where to move to as I am not sure where would be best?

File: app/questionnaire/rules.py
Comment (iwootten) [inline app/questionnaire/rules.py]: This could do with being better named, due to a lot of other similarly named methods. Maybe get_answer_on_routing_path?

File: app/libs/utils.py
Comment (iwootten) [reply app/libs/utils.py]: I've been looking at this and think it might be clearer if it only returned the answer tuple to lookup.

File: app/data_model/progress_store.py
Comment (LJBabbage) [inline app/data_model/progress_store.py]: Really minor, but for continuity as this is effectively the origin of the name section_key and section_progress, the comment could reflect that.

File: app/questionnaire/path_finder.py
Comment (LJBabbage) [inline app/questionnaire/path_finder.py]: Nothing major again, just a slight dislike of the name 'for_list', not sure its represents what it is

File: app/questionnaire/path_finder.py
Comment (LJBabbage) [inline app/questionnaire/path_finder.py]: As above, just might make line 137 and 141 a bit easier if looked at out of context

File: app/questionnaire/router.py
Comment (LJBabbage) [inline app/questionnaire/router.py]: As above

File: app/questionnaire/rules.py
Comment (LJBabbage) [inline app/questionnaire/rules.py]: Minor typo

File: app/views/contexts/hub_context.py
Comment (LJBabbage) [inline app/views/contexts/hub_context.py]: As above

File: app/views/contexts/hub_context.py
Comment (LJBabbage) [inline app/views/contexts/hub_context.py:114]: Pretty sure its just me, but I don't really understand what rows are in this context. Looked at the function and it just looks like a dict of stuff. Is this for the frontend table of repeating elements ?

File: data-source/json/test_repeating_sections_with_hub_and_spoke.json
Comment (LJBabbage) [inline data-source/json/test_repeating_sections_with_hub_and_spoke.json]: Just showing I am still checking this monster PR, missing the '?'

**Review by LJBabbage — CHANGES_REQUESTED** [review]
Great work, specially with the tests. Works with your new schema update and a few other random ones to test regression. Just a few minor points.

File: app/data_model/progress_store.py
Comment (JamesGardiner) [reply app/data_model/progress_store.py]: Does it even need this comment? It should be obvious enough from the code what this is doing :wink:

File: app/data_model/progress.py
Comment (JamesGardiner) [inline app/data_model/progress.py]: Imports aren't consistent... Either use one line per root import module or split them. Not both.

File: app/data_model/progress.py
Comment (JamesGardiner) [inline app/data_model/progress.py]: Any reason you want to use Dict here over Mapping? Mapping is usually preferred.

File: app/data_model/progress.py
Comment (JamesGardiner) [inline app/data_model/progress.py]: Same question about dict vs mapping

File: app/data_model/progress_store.py
Comment (JamesGardiner) [inline app/data_model/progress_store.py:88]: Could we make this EAFP rather than LBYL?

Comment (JamesGardiner) [reply app/data_model/progress_store.py:88]: Same for the rest of this module.

File: app/data_model/progress_store.py
Comment (JamesGardiner) [inline app/data_model/progress_store.py]:
```Python
section_keys_to_delete = [(section_key[0], section_key[1]) for section_key in self._progress if section_key[1] == list_item_id]
```

File: app/libs/utils.py
Comment (JamesGardiner) [reply app/libs/utils.py]: I agree with Ian. This conflates getting things from the answer store with resolving the id of an answer.

File: app/questionnaire/path_finder.py
Comment (JamesGardiner) [reply app/questionnaire/path_finder.py]: I've been thinking this for a while but can't conjure up anything more meaningful - any suggestions @LJBabbage ?

Comment (JamesGardiner) [inline app/questionnaire/path_finder.py]: `get_repeating_list_for_section` should probably raise a custom exception to be handled when no repeating list for a given section id is found. This could then be a try/except block. Something like `NoRepeatingSectionFound`

File: app/questionnaire/path_finder.py
Comment (JamesGardiner) [reply app/questionnaire/path_finder.py:89]: Any reason for dropping the type hint on the section_id arg?

File: app/questionnaire/questionnaire_schema.py
Comment (JamesGardiner) [inline app/questionnaire/questionnaire_schema.py:41]: I think these should start raising exceptions rather than None when things aren't found. It makes the calling code easier to write as EAFP and feels more pythonic to me. Happy to take a view on this from the team though.

File: app/views/handlers/block_factory.py
Comment (JamesGardiner) [inline app/views/handlers/block_factory.py]: Can we make this a try except block using KeyError for the dict access?

**Review by JamesGardiner — CHANGES_REQUESTED** [review]

**Comment by JamesGardiner** [comment]
Can we keep consistency for the use of Dict vs Mapping as a type hint? I prefer Mapping, as it is more flexible than dict in that  it accepts any object that implements the mapping protocol

File: app/questionnaire/path_finder.py
Comment (LJBabbage) [reply app/questionnaire/path_finder.py]: Yeah, not the easiest, 'section_repeating_list' is a mouthful, but the best I have at the minute.

File: app/data_model/progress_store.py
Comment (LJBabbage) [reply app/data_model/progress_store.py]: True, but if coming in completely cold, these little things just make it a little easier to understand the code. Dealers choice I would say

File: app/libs/utils.py
Comment (MebinAbraham) [reply app/libs/utils.py]: This method has been removed. Is partly found in `rules.py`, not sure where it should live.

File: app/questionnaire/rules.py
Comment (MebinAbraham) [reply app/questionnaire/rules.py]: I've gone with `get_answer_value`.

File: app/questionnaire/path_finder.py
Comment (JamesGardiner) [reply app/questionnaire/path_finder.py]: Ah the problem with that is that for_list is used elsewhere for other actions, such as in relationship collectors

File: app/data_model/progress_store.py
Comment (MebinAbraham) [reply app/data_model/progress_store.py]: I've added extra context to this, let me know if it is overkill.

File: app/questionnaire/path_finder.py
Comment (MebinAbraham) [reply app/questionnaire/path_finder.py]: So I used the name `for_list` since that is what we call it in the schema.
I've gone with `repeating_list` in the end.

File: app/questionnaire/path_finder.py
Comment (MebinAbraham) [reply app/questionnaire/path_finder.py]: https://github.com/ONSdigital/eq-survey-runner/pull/2248#discussion_r313362102

File: app/questionnaire/router.py
Comment (MebinAbraham) [reply app/questionnaire/router.py]: https://github.com/ONSdigital/eq-survey-runner/pull/2248#discussion_r313362102

File: app/questionnaire/rules.py
Comment (MebinAbraham) [reply app/questionnaire/rules.py]: Just making sure you are actually reviewing 😄 
Fixed.

File: app/views/contexts/hub_context.py
Comment (MebinAbraham) [reply app/views/contexts/hub_context.py]: https://github.com/ONSdigital/eq-survey-runner/pull/2248#discussion_r313362102

File: app/views/contexts/hub_context.py
Comment (MebinAbraham) [reply app/views/contexts/hub_context.py:114]: Yes, this is to create the context for the rows displayed on the hub.

File: app/data_model/progress.py
Comment (MebinAbraham) [reply app/data_model/progress.py]: Not enforced by linting but fixed.

File: app/data_model/progress.py
Comment (MebinAbraham) [reply app/data_model/progress.py]: I've kept it consistent with Answer.py since it is a literal dict. I can change to Mapping but almost all places we use this is a literal dict?

File: app/questionnaire/path_finder.py
Comment (MebinAbraham) [reply app/questionnaire/path_finder.py:89]: Reintroduced.

File: app/data_model/progress.py
Comment (MebinAbraham) [reply app/data_model/progress.py]: https://github.com/ONSdigital/eq-survey-runner/pull/2248#discussion_r313364655

File: app/data_model/progress_store.py
Comment (MebinAbraham) [reply app/data_model/progress_store.py:88]: To be honest, I personally prefer LBYL and keeps it consistent with our current codebase. I think we should make a decision as a team on an approach because the `progress_store` went through peer reviews and was approved with the LBYL approach. 

I think changing this to EAFP is out of scope unless the team decides that is what we should go with from now on.

File: app/data_model/progress_store.py
Comment (MebinAbraham) [reply app/data_model/progress_store.py]: Done with a slight change.

File: app/questionnaire/path_finder.py
Comment (MebinAbraham) [reply app/questionnaire/path_finder.py]: As mentioned in https://github.com/ONSdigital/eq-survey-runner/pull/2248#discussion_r313374541. To me, it seems unnecessary to raise an exception for this, I think this is fine unless the team wants to change the approach to EAFP.

File: app/questionnaire/questionnaire_schema.py
Comment (MebinAbraham) [reply app/questionnaire/questionnaire_schema.py:41]: The same comment as previously, the team should make a choice. If we do EAFP then everything should be doing that even `QuestionnaireSchema` otherwise it would just be EAFP that calls code that does LBYL etc.

File: app/views/handlers/block_factory.py
Comment (MebinAbraham) [reply app/views/handlers/block_factory.py]: Same as my previous comments. The code above mine is also doing `.get`, think we should be consistent.

**Comment by MebinAbraham** [comment]
> Can we keep consistency for the use of Dict vs Mapping as a type hint? I prefer Mapping, as it is more flexible than dict in that it accepts any object that implements the mapping protocol

I am happy to change to Mapping if that is deemed the preference. 

From online:

> typing.Dict should be used to indicate a literal dict type with support for element type hinting i.e. Dict[bytes, str]
> 
> typing.Mapping is an object which defines the __getitem__,__len__,__iter__ magic methods

Most of our cases are literal dict?

**Comment by JamesGardiner** [comment]
I still think go with mapping, should we ever define our own mapping object there will be no need to update the type hints

File: app/views/contexts/hub_context.py
Comment (LJBabbage) [reply app/views/contexts/hub_context.py:114]: Thank you

**Review by LJBabbage — APPROVED** [review]

File: app/data_model/progress.py
Comment (MebinAbraham) [reply app/data_model/progress.py]: This is now `MutableMapping`

File: app/data_model/progress_store.py
Comment (iwootten) [inline app/data_model/progress_store.py]: Is this actually a section or a representation of progress?

**Review by JamesGardiner — APPROVED** [review]

File: app/data_model/progress_store.py
Comment (iwootten) [inline app/data_model/progress_store.py]: As above, are these sections? Could they be named better if not?

File: app/questionnaire/path_finder.py
Comment (iwootten) [inline app/questionnaire/path_finder.py]:
```suggestion
        sections = sections or self.schema.get_sections()
```

File: app/questionnaire/path_finder.py
Comment (iwootten) [inline app/questionnaire/path_finder.py]:
```suggestion
        for section in sections:
```

File: app/data_model/progress_store.py
Comment (bitdivision) [inline app/data_model/progress_store.py]: We should review whether this needs to be mutable. May be better done as a seperate PR.

File: app/data_model/progress_store.py
Comment (bitdivision) [inline app/data_model/progress_store.py]: Why remove the type hints?

File: app/questionnaire/questionnaire_store_updater.py
Comment (iwootten) [inline app/questionnaire/questionnaire_store_updater.py]: This could be simplified to:
```suggestion
        location = location or self._current_location
        
        self._progress_store.update_section_status(
            section_status, section_id=location.section_id, list_item_id=location.list_item_id
        )
```

File: app/questionnaire/path_finder.py
Comment (bitdivision) [inline app/questionnaire/path_finder.py]: When is this sections acgument used?

File: tests/functional/spec/features/repeating_sections/repeating_sections_with_hub_and_spoke.js
Comment (iwootten) [inline tests/functional/spec/features/repeating_sections/repeating_sections_with_hub_and_spoke.js]:
```suggestion
    beforeEach('Navigate to the Hub', function () {
```

File: app/data_model/progress_store.py
Comment (MebinAbraham) [reply app/data_model/progress_store.py]: As discussed, this is now `section_progress`

File: app/data_model/progress_store.py
Comment (MebinAbraham) [reply app/data_model/progress_store.py]: As discussed, this is now `section_progress in section_progress_list`

File: app/questionnaire/path_finder.py
Comment (MebinAbraham) [reply app/questionnaire/path_finder.py]: Done

File: app/questionnaire/path_finder.py
Comment (MebinAbraham) [reply app/questionnaire/path_finder.py]: Done

File: app/questionnaire/questionnaire_store_updater.py
Comment (MebinAbraham) [reply app/questionnaire/questionnaire_store_updater.py]: Done

**Review by iwootten — APPROVED** [review]

File: app/questionnaire/questionnaire_store_updater.py
Comment (bitdivision) [inline app/questionnaire/questionnaire_store_updater.py]: Should the arguments of this match the questionnaire_store method?

**Review by bitdivision — CHANGES_REQUESTED** [review]

File: app/questionnaire/router.py
Comment (bitdivision) [inline app/questionnaire/router.py]: Rather than passing in a tuple here, should progress_store have an is_section_completed method which we can pass a section_id and list_item_id?

File: app/data_model/progress_store.py
Comment (MebinAbraham) [reply app/data_model/progress_store.py]: Have changed where I can to be `Mapping` but yes needs to be reviewed as a separate task.

File: app/views/handlers/block_factory.py
Comment (bitdivision) [inline app/views/handlers/block_factory.py]: Why is this a kwarg?

File: app/views/handlers/block.py
Comment (bitdivision) [inline app/views/handlers/block.py:33]: Can this just expose current_location rather than making this a property?

File: app/questionnaire/questionnaire_store_updater.py
Comment (MebinAbraham) [reply app/questionnaire/questionnaire_store_updater.py]: Maybe, felt weird to pass in location to update a section status, so got questionnaire store update to handle it. Could name it `update_section_status_for_location`?

File: app/questionnaire/router.py
Comment (MebinAbraham) [reply app/questionnaire/router.py]: Done with the section_key, or are you suggesting specifically to do it with section_id, list_item_id?

Comment (MebinAbraham) [reply app/questionnaire/router.py]: I'll do it with `section_id` and `list_item_id` to keep it consistent with the rest of the module.

File: app/data_model/progress_store.py
Comment (MebinAbraham) [reply app/data_model/progress_store.py]: Not sure, added type hints to rest of the module.

File: app/questionnaire/path_finder.py
Comment (MebinAbraham) [reply app/questionnaire/path_finder.py]: Just making sure you are reviewing properly. Good spot 😁  Removed.

File: app/questionnaire/router.py
Comment (bitdivision) [reply app/questionnaire/router.py]: Yeah. To me `section_key` is an implementation detail of the store, we shouldn't be exposing it if possible.

File: app/views/handlers/block.py
Comment (MebinAbraham) [reply app/views/handlers/block.py:33]: Can do, only did this to prevent it from being mutated?

File: app/views/handlers/block_factory.py
Comment (MebinAbraham) [reply app/views/handlers/block_factory.py]: Removed kwarg and made it an arg with a default.

File: app/views/handlers/block.py
Comment (bitdivision) [reply app/views/handlers/block.py:33]: Yep, fair enough

File: app/questionnaire/questionnaire_store_updater.py
Comment (MebinAbraham) [reply app/questionnaire/questionnaire_store_updater.py]: Have changed it to match the progress store method.

**Review by bitdivision — APPROVED** [review]
Nice work Mebin

**Pull request merged**
